src/contracts/libraries/GPv2Signing.sol

```diff
@@ -167,6 +169,99 @@ library GPv2Signing {
         setFreeMemoryPointer(freeMemoryPointer);
     }
 
+    /// @dev Verifies the input calldata as an EIP-1271 contract signature and
+    /// returns the address of the signer.
+    ///
+    /// The encoded signature tightly packs the following struct:
+    ///
+    /// ```
+    /// struct EncodedErc1271Signature {
+    ///     address verifier;
+    ///     uint16 signatureLength;
+    ///     bytes signature;
+    /// }
+    /// ```
+    ///
+    /// All entries are tightly packed together in this order in the encoded
+    /// calldata. `signatureLength` encodes the length of the signature bytes
+    /// in bytes. Example:
+    ///
+    /// input:    0x73c14081446bd1e4eb165250e826e80c5a523783000a00010203040506070809
+    /// decoding:   [..............verifier................][sL][.......data.......]
+    /// stride:                                          20   2         10 (0x000a)
+    ///
+    /// This function enforces that the encoded data stores enough bytes to
+    /// cover the full length of the decoded interaction.
+    ///
+    /// The size of `dataLength` limits the maximum calldata that can be used in
+    /// a signature to 2**16 ≈ 65 kB.
+    function recoverEip1271Signer(
+        bytes calldata encodedSignature,
+        bytes32 domainSeparator,
+        bytes32 orderDigest
+    ) internal view returns (address owner, bytes calldata remainingCalldata) {
+        uint256 signatureLength;
+        bytes calldata signature;
+
+        // NOTE: Use assembly to efficiently decode signature data.
+        // If reading calldata out of bound, the extra bytes are set to zero.
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            // owner = address(encodedSignature[0:20])
+            owner := shr(96, calldataload(encodedSignature.offset))
+            // signatureLength = uint256(encodedSignature[20:22])
+            signatureLength := shr(
+                240,
+                calldataload(add(encodedSignature.offset, 20))
+            )
+        }
+
+        // Safety: dataLength fits a uint16 by construction, no overflow is
+        // possible.
+        uint256 usedCalldataStride = 20 + 2 + signatureLength;
+        require(
+            encodedSignature.length >= usedCalldataStride,
+            "GPv2: invalid signature encoding"
```

Comment:
It's hard to be very specific:
```
warning  Error message for require is too long
```
On the other hand, I think this message might already be misleading: the error is in the decoding, so it might be that the trade decoding was incorrect, which caused the chosen length bit was the wrong one and triggered this issue. We use `GPv2: invalid encoding` for the other two signature types, I think it'd be reasonable to use the same here.

Reply:
You could use `signature too long` or something. But I'm also OK with how it is right now as we use `invalid X encoding` for other decoding errors.

IMO revert messages are better served as "error codes" than descriptive messages.

Reply:
Should I then differentiate the error messages for a failed signature decoding (not verification) in ECDSA signature?
The messages currently are kind of bad:
```
ECDSA   -> GPv2: invalid encoding
EIP1271 -> GPv2: invalid signature encoding
```
My previous proposal would be to make both "GPv2: invalid encoding". In general I expect that if a transaction fails there, the error is likely outside of the signature itself. It might still make sense to split, maybe as:
```
ECDSA   -> GPv2: invalid encoding, ecdsa
EIP1271 -> GPv2: invalid encoding, eip1271
```
This helps for testing, since we can catch failures with `revertsWith("invalid encoding")`. However it might be that what fails is the actual signature decoding, then this might be more appropriate:
```
ECDSA   -> GPv2: signature too long, ecdsa
EIP1271 -> GPv2: signature too long, eip1271
```
WDY prefer?

Reply:
Golfing the error message! Ha ha.